mapcss/MapCSS.g4

```diff
@@ -77,7 +78,7 @@ OP_PLUS: '+';
 OP_MINUS: '-';
 OP_NOT: '!'; // NOTE: boolean not -> !(expr)
 
-SET: ('s' | 'S') ('e' | 'E') ('t' | 'T');
+SET: ('s' | 'S') ('e' | 'E') ('t' | 'T') (' ' | '\t') ?;
```

Comment:
No, parsing space is not the right option.
It probably an issue with some missing space in the output printing.

Reply:
`set` itself is never defined to contain a space. 
Already in the first (not-generated-by-ANTLR) file, in the function `exitDeclaration`, it looks like this without the patch:
`{'type': 'declaration', 'text': 'setarea_yes_autofix', 'set': 'area_yes_autofix', 'property': None, 'value': None}`
while, with the patch, it looks like this:
`{'type': 'declaration', 'text': 'set area_yes_autofix', 'set': 'area_yes_autofix', 'property': None, 'value': None}`

The `text` property is what's printed out in `to_p` in mapcss2osmose:
`selectors_text = "# " + "\n# ".join(map(lambda s: s['text'], t['selectors']))`
Adding exceptions for `set` there is probably going to be very ugly.

Spaces are just discarded everywhere:
`WS:           (' ' | '\t' | '\n' | '\r' | '\f') -> channel(HIDDEN);`
(That's also why for example [this](https://github.com/osm-fr/osmose-backend/blob/master/plugins/Colour.validator.mapcss#L70) many-space-containing mapcss rule is [changed to](https://github.com/osm-fr/osmose-backend/blob/master/plugins/Colour.py#L99) `# *[colour:arrow][colour:arrow=~/^([0-9A-Fa-f]{3}|[0-9A-Fa-f]{6})$/]`, without spaces: spaces are dropped unless explicitly captured)

So either we have to explicitly capture it (like in the quoted strings (used in i.e. `[a="b c"]`) where spaces must also be retained, e.g.:
`DQUOTED_STRING: '"' (' ' | '!' | '#'..'[' | ']'..'~' | '°' | UNICODE | EDQUOTE | EBACKSLASH)* '"';`
or we would have to change exitDeclaration to get a special case for set, for example:
```
    def exitDeclaration(self, ctx:MapCSSParser.DeclarationContext):
        v = self.stack.pop()
        self.declarations.append({
            'type': 'declaration',
            'text': ctx.getText(),
            'set': ctx.cssident() and ctx.cssident().getText(),
            'property': ctx.declaration_property() and ctx.declaration_property().getText(),
            'value': v['single_value'] or v['functionExpression']
        })
        if self.declarations[-1]["set"]:
            self.declarations[-1]["text"] = self.declarations[-1]["text"].replace("set", "set ", 1)
```

I would argue it's neater to capture it than to have such special cases. Would you prefer the latter?

Reply:
There also the same issue with `import` (but not used/implemented).
I look at how to implement it. I think it may be something like this (not tested)
```diff
         self.declarations.append({                                                                                       
             'type': 'declaration',                                                                                       
-             'text': ctx.getText(),                                                                                       
+             'text': ctx.getText() + (" " if self.declarations[-1]["set"] else ""),                                                                                       
             'set': ctx.cssident() and ctx.cssident().getText(),                                                          
             'property': ctx.declaration_property() and ctx.declaration_property().getText(),                             
             'value': v['single_value'] or v['functionExpression']                                                        
         })  
```

Reply:
> I look at how to implement it.

Thank you!

Be aware that `self.declarations[-1]["set"]` is only defined the line after you use it in your example code, so I guess you'd need to create a variable like `set_value`, as below (also untested)
```
         set_value = ctx.cssident() and ctx.cssident().getText()
         self.declarations.append({                                                                                       
             'type': 'declaration',                                                                                       
             'text': ctx.getText() + (" " if set_value else ""),                                                                                       
             'set': set_value,                                                          
             'property': ctx.declaration_property() and ctx.declaration_property().getText(),                             
             'value': v['single_value'] or v['functionExpression']                                                        
         })  
```

Reply:
> I look at how to implement it.



In case you need inspiration, you could consider 
`'text': re.sub(r'^[sS][eE][tT]', 'set ', ctx.getText()),`

Note that regular `.replace('set', 'set ', 1)` wouldn't work because:
1. the definition of `SET` also matches SeT, sET, ..., 
2. `declaration` (and thus the text property) could, besides set declarations, also be i.e. `declaration_property COLON declaration_value` where `declaration_value` may contain `set` in a word, i.e. 
```
way[surface=cobblestone] {
  throwWarning: 'Prefer sett or unhewn_cobblestone over cobblestone';
}
```
would result in `ctx.getText()` outputting `throwWarning:'Prefer sett or unhewn_cobblestone over cobblestone'` (so including the word `set` in `sett`, where you shouldn't add a space)